File: cours-04/liste-trie.py
import math

class MyArray:
    innerArray = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    size = 0
    
    def insert(self, n):
        if(self.size == 0):
            self.innerArray[0] = n
        else:
            for i in reversed(range(0, self.size)):
                if(self.innerArray[i] > n):
                    self.innerArray[i + 1] = self.innerArray[i]
                else:
                    self.innerArray[i + 1] = n
                    break
        
        self.size += 1
        
    def search(self, n):
        lower = 0
        higher = self.size - 1
        
        while lower <= higher:
            middle = math.ceil(lower + (higher - lower) / 2)
            
            if n == self.innerArray[middle]:
                return middle;
            if self.innerArray[middle] > n:
                higher = middle - 1
            if self.innerArray[middle] < n:
                lower = middle + 1
        
        return -1
        

intArray = MyArray()
intArray.insert(15)
intArray.insert(42)
intArray.insert(20)
intArray.insert(30)
intArray.insert(70)
intArray.insert(58)
intArray.insert(40)
intArray.insert(55)

print(intArray.search(70))

File: cours-04/liste-trie.java
class MyArray {
    private int[] innerArray = new int[100];
    private int size = 0;

    public int get(int index) {
        return innerArray[index];
    }

    // Big(O): n
    public void insert(int n) {
        if(size == 0) {
            innerArray[0] = n;
        } else {
            for(int i = size - 1; i >= 0; i--) {
                if(innerArray[i] > n) {
                    innerArray[i + 1] = innerArray[i];
                } else {
                    innerArray[i + 1] = n;
                    break;
                }
            }
        }

        size++;
    }

    // Big(O): n
    public int search(int n) {
        for(int i = 0; i < size; i++) {
            if(innerArray[i] == n)
                return i;
        }

        return -1;
    }

    // Big(O): log(n)
    public int searchBinary(int n) {
        int lower = 0;
        int higher = size - 1;
        int middle;

        do {
            middle = (int)Math.ceil(lower + (higher - lower) / 2.0);
            if(innerArray[middle] > n) {
                higher = middle;
            }
            if(innerArray[middle] < n) {
                lower = middle;
            }
        } while(n != innerArray[middle]
                && lower != higher);

        if(innerArray[middle] == n)
            return middle;
        else
            return -1;
    }
}

void main() {
    MyArray intArray = new MyArray();
    intArray.insert(15);
    intArray.insert(42);
    intArray.insert(20);
    intArray.insert(30);
    intArray.insert(70);
    intArray.insert(58);
    intArray.insert(40);
    intArray.insert(55);

    System.out.println(intArray.searchBinary(58));
}